Changing the ThingType of existing Things

sjka opened the pull request:
This PR should allow bindings to change the ThingType of a Thing after it was created.

Therefore, it comprises the following changes:
* Allowing to create ThingUIDs without specifying the ThingType
* A Thing's ThingType should not be read from the ThingUID anymore, but from the Thing directly
* A callback method was introduced, allowing the ThingHandler to initiate the ThingType migration

I have kept several commits for now, in order keeps things easier to read. 
The actual migration happens in the ThingManager.

The changing of the ThingType happens under the following assumptions:
* The Thing object remains
* The ThingHandler gets shut down properly, a new one will be registered
* The Thing transitions through a new initialization lifecycle
* The Channels will be dropped and pre-populated according to the new ThingType

These assumptions are tested in ChangeThingTypeOSGiTest

Yet known open topics
* Are the above assumptions valid for your use-cases?
* The old channels might have some configurations/properties - what should happen with them?
* JavaDoc has not been updated everywhere
* Some polishing and additional test cases are yet missing
* Squashing of the commits

Feel free to have a look and point out any conceptional concerns you might have.

kaikreuzer commented:
Hey @cdjackson, Could you have a brief look if this about meets your needs?

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingHandlerCallback.java:
```diff
@@ -49,10 +51,12 @@
      */
     void thingUpdated(Thing thing);
 
+    void changeThingType(Thing thing, ThingTypeUID thingTypeUID, Configuration configuration);
```
some JavaDoc would be nice :-)

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java:
```diff
@@ -236,6 +242,11 @@ public Thing run() {
 
+        @Override
+        public void changeThingType(Thing thing, ThingTypeUID thingTypeUID, Configuration configuration) {
+            migrateThingType(thing, thingTypeUID, configuration);
```
Why introducing an extra method here that is only called from this place? You could put the code right in here!

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java:
```diff
@@ -661,4 +672,56 @@ private void setThingStatus(Thing thing, ThingStatusInfo thingStatusInfo) {
+        }
+
+        // Remove the links
+        thingLinkManager.thingRemoved(thing);
```
since you are doing a thingUpdated() call in the end, you should not need to manually deal with the links here.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java:
```diff
@@ -661,4 +672,56 @@ private void setThingStatus(Thing thing, ThingStatusInfo thingStatusInfo) {
+        ((ThingImpl) thing).setThingTypeUID(thingTypeUID);
+
+        // Establish links
+        thingLinkManager.thingAdded(thing);
```
since you are doing a thingUpdated() call in the end, you should not need to manually deal with the links here.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java:
```diff
@@ -661,4 +672,56 @@ private void setThingStatus(Thing thing, ThingStatusInfo thingStatusInfo) {
+        // Register the new Handler
+        registerHandler(thing);
+
+        thingUpdated(thing, null); // TODO needed? Assumption: Yes, at least for Thing-Item-Links
```
Needed, if you remove the calls above.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingConfigDescriptionProvider.java:
```diff
@@ -57,8 +57,8 @@ public ConfigDescription getConfigDescription(URI uri, Locale locale) {
         // First, get the thing type so we get the generic config descriptions
-        ThingUID thingUID = new ThingUID(uri.getSchemeSpecificPart());
-        ThingType thingType = thingTypeRegistry.getThingType(thingUID.getThingTypeUID());
+        ThingTypeUID thingTypeUID = new ThingTypeUID(uri.getSchemeSpecificPart());
```
Why is `uri.getSchemeSpecificPart()` all of a sudden a thing type UID and not anymore a thing UID? Note that we are within the ThingConfigDescriptionProvider, which deals with uris of Things.

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingFactoryHelper.java:
```diff
@@ -0,0 +1,158 @@
+package org.eclipse.smarthome.core.thing.internal;
```
please add license header

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingFactoryHelper.java:
```diff
@@ -0,0 +1,158 @@
+
+import com.google.common.collect.Lists;
+
+public class ThingFactoryHelper {
```
please add javadoc

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingFactoryHelper.java:
```diff
@@ -0,0 +1,158 @@
+
+    private static Logger logger = LoggerFactory.getLogger(ThingFactory.class);
+
+    public static List<Channel> createChannels(ThingType thingType, ThingUID thingUID,
```
please add javadoc

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingFactoryHelper.java:
```diff
@@ -0,0 +1,158 @@
+        return channels;
+    }
+
+    private static Channel createChannel(ChannelDefinition channelDefinition, ThingUID thingUID,
```
please add javadoc

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingFactoryHelper.java:
```diff
@@ -0,0 +1,158 @@
+        return channel;
+    }
+
+    public static Object getDefaultValueAsCorrectType(Type parameterType, String defaultValue) {
```
please add javadoc

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingFactoryHelper.java:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    public static void applyDefaultConfiguration(Configuration configuration, ThingType thingType,
```
please add javadoc

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingImpl.java:
```diff
@@ -81,13 +81,26 @@ public ThingImpl(ThingTypeUID thingTypeUID, String thingId) throws IllegalArgume
+    @Deprecated
     public ThingImpl(ThingUID thingUID) throws IllegalArgumentException {
         this.uid = thingUID;
         this.thingTypeUID = new ThingTypeUID(thingUID.getBindingId(), thingUID.getThingTypeId());
```
Shouldn't this now throw an IllegalArgumentException in case the thingTypeId is empty?

kaikreuzer commented on bundles/config/org.eclipse.smarthome.config.discovery/src/main/java/org/eclipse/smarthome/config/discovery/internal/DiscoveryResultImpl.java:
```diff
@@ -52,10 +53,36 @@
+     * @param thingTypeUID the {@link ThingTypeUID}
+     * @param thingUID
+     *            the Thing UID to be set (must not be null). If a {@code Thing} disappears and is discovered again, the
+     *            same {@code Thing} ID
```
why are here line breaks in the text?

kaikreuzer commented:
lgtm in general, thanks!

cdjackson commented:
@sjka did you confirm that discovery properties are correctly handled still? In the discovery result, I pass in some properties (address) and this is then extracted in the thing.initialize() method. This used to work, but now properties appears to be empty.
I traced this through and it looked ok in the discovery builder, but I suspect it must be on the discovery builder saving to mapdb side as when I swapped back to the old code, the inbox still didn't show the properties until I ran discovery again...

kaikreuzer commented:
Well, at least currently the migration code calls updateThing() in https://github.com/eclipse/smarthome/pull/768/files#diff-24d71ac793f6de6c8557ea06292497ddR708 and this means that the changes on the Thing have to be persisted - which is obviously not possible if it is defined in a file.

kaikreuzer commented:
@sjka Any chance to incorporate the feedback comments any time soon?

kaikreuzer commented:
@sjka Any news here?

sjka commented:
Sorry for the huge delay...
I have updated the PR with your comments. 
Nevertheless, I can't really figure out the observations that @cdjackson made and don't really see how this could be related. I rebased my change though - maybe that caused any side-effects? Could you please help me to reproduce the issue?

cdjackson commented:
@sjka the problem I was having seems to be gone - not sure if it was an issue here or if it was fixed with the rebase...  I'll try and spend some more time over the next week to play with this.

Thanks (and I hear congratulations are in order :) ).

kaikreuzer commented:
Thanks guys, so overall it looks good then. I'll also do some further testing and if @cdjackson does not come back with any problems by the end of the week, we should merge it.

Before we can merge, the Paper UI also needs to be adapted though. Here is what Dennis left me as info about the necessary changes:
"Ganz oben in dieser Datei (https://github.com/eclipse/smarthome/blob/master/extensions/ui/org.eclipse.smarthome.ui.paper/web/js/controllers.configuration.js) ist die Methode "getThingTypeUID". Die wird an einigen Stellen in dem configuration-Controller verwendet. Das Refactoring wäre im wesentlichen zuerst den REST-Request für das Thing abzuwarten, um dann den Request für den ThingType anhand der ThingType UID aus dem Thing zu machen. Bisher geschieht das aus Performance-Gründen parallel, da ich noch davon ausgegangen bin, dass die ThingUID halt immer die ThingTypeUID enthält."

I'll ask @aounhaider1 to take care of that in the course of this week.

cdjackson commented:
So, I'm not having 100% success...

Firstly, I often (not always) get the following exception -:

    2016-01-25 22:44:05.355 [ERROR] [.c.thing.internal.ThingManager:494  ] - Exception occured while calling thing updated at ThingHandler 'org.eclipse.smarthome.binding.ble.handler.YeelightBlueThingHandler@5b24574e: nulljava.util.concurrent.TimeoutException: null
	at java.util.concurrent.FutureTask.get(FutureTask.java:205)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:157)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.thingUpdated(ThingManager.java:484)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$1.changeThingType(ThingManager.java:280)
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.changeThingType(BaseThingHandler.java:589)
	at org.eclipse.smarthome.binding.ble.handler.BleGenericThingHandler.discoveryComplete(BleGenericThingHandler.java:112)
	at org.eclipse.smarthome.binding.ble.handler.BleGenericThingHandler.handleDescriptorRead(BleGenericThingHandler.java:91)
	at org.eclipse.smarthome.binding.ble.handler.BleBaseThingHandler$GattCallback.onDescriptorRead(BleBaseThingHandler.java:374)
	at org.eclipse.smarthome.io.transport.bluetooth.bluez.BluezBluetoothGattDescriptor$ReadThread.run(BluezBluetoothGattDescriptor.java:198)

I'm also unclear what happens after the type is changed. If I look in the thing, it does seem that it's got the new type (when viewed inside the *changeThingType* method), but I'm still seeing debug messages logged from the old thing handler.  If I restart OH after the change, it is still instantiated with the original type.

A question on it's use - is there any restrictions on where I can all the changeThingType method? Currently, I call it from within the original thing handler (since this is where I'm doing all the discovery)...

As an overview, after initial discover, I put a 'generic' thing type into the inbox. Once this is instantiated, I do discovery in my 'genericHandler' and once I know what the device is, I then call the changeThingType method (and get the above exception).

The following logging (I think?) seems to reflect that the type isn't changed (although I've not tried to follow the logging through, so you may disagree) -:

    2016-01-25 23:16:16.861 [DEBUG] [.c.thing.internal.ThingManager:579  ] - Calling unregisterHandler handler for thing 'ble:generic:544a16204bae' at 'org.eclipse.smarthome.binding.ble.internal.BleHandlerFactory@19c9f7bf'.
    2016-01-25 23:16:16.879 [DEBUG] [.c.thing.internal.ThingManager:109  ] - Thing handler for thing 'ble:generic:544a16204bae' removed.
    2016-01-25 23:16:16.880 [DEBUG] [.c.thing.internal.ThingManager:322  ] - Unassigning handler for thing 'ble:generic:544a16204bae' and setting status to UNINITIALIZED.
    2016-01-25 23:16:16.880 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'ble:generic:544a16204bae' updated: UNINITIALIZED (HANDLER_MISSING_ERROR)
    2016-01-25 23:16:16.881 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'ble:generic:544a16204bae' changed from OFFLINE to UNINITIALIZED (HANDLER_MISSING_ERROR)
    2016-01-25 23:16:21.890 [ERROR] [.c.thing.internal.ThingManager:589  ] - Exception occured while calling thing handler factory 'org.eclipse.smarthome.binding.ble.internal.BleHandlerFactory@19c9f7bf': nulljava.util.concurrent.TimeoutException: null
	at java.util.concurrent.FutureTask.get(FutureTask.java:205)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:157)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.unregisterHandler(ThingManager.java:581)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.access$9(ThingManager.java:578)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$1.changeThingType(ThingManager.java:253)
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.changeThingType(BaseThingHandler.java:589)
	at org.eclipse.smarthome.binding.ble.handler.BleGenericThingHandler.discoveryComplete(BleGenericThingHandler.java:112)
	at org.eclipse.smarthome.binding.ble.handler.BleGenericThingHandler.handleDescriptorRead(BleGenericThingHandler.java:91)
	at org.eclipse.smarthome.binding.ble.handler.BleBaseThingHandler$GattCallback.onDescriptorRead(BleBaseThingHandler.java:374)
    2016-01-25 23:16:21.902 [DEBUG] [.c.thing.internal.ThingManager:543  ] - Calling registerHandler handler for thing 'ble:generic:544a16204bae' at 'org.eclipse.smarthome.binding.ble.internal.BleHandlerFactory@19c9f7bf'.
    2016-01-25 23:16:21.905 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'ble:generic:544a16204bae' updated: INITIALIZING
    2016-01-25 23:16:21.908 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'ble:generic:544a16204bae' changed from UNINITIALIZED (HANDLER_MISSING_ERROR) to INITIALIZING
    2016-01-25 23:16:26.911 [INFO ] [ome.event.ThingStatusInfoEvent:43   ] - 'ble:generic:544a16204bae' updated: UNINITIALIZED (HANDLER_INITIALIZING_ERROR)
    2016-01-25 23:16:26.915 [INFO ] [nt.ThingStatusInfoChangedEvent:43   ] - 'ble:generic:544a16204bae' changed from INITIALIZING to UNINITIALIZED (HANDLER_INITIALIZING_ERROR)
    2016-01-25 23:16:26.914 [WARN ] [.c.thing.internal.ThingManager:567  ] - Registering handler for thing 'ble:generic:544a16204bae' takes more than 5000ms.
    2016-01-25 23:16:26.918 [DEBUG] [.c.t.internal.ThingLinkManager:376  ] - Removing linked group item from thing 'ble:generic:544a16204bae'.
    2016-01-25 23:16:26.922 [DEBUG] [.c.t.internal.ThingLinkManager:328  ] - Assigning linked group item 'ble_generic_544a16204bae' to thing 'ble:generic:544a16204bae'.

It seems to show the generic handler being removed, then added again?

This is the code -:

    // Change the thing type
    Configuration configuration = new Configuration();
    changeThingType(BleBindingConstants.THING_TYPE_YEELIGHT_BLUE, configuration);

sjka commented:
@cdjackson Thanks for that observation - there indeed was a synchronization issue leading to a dead-lock, which I fixed now. So it should not matter from where you call it. Usually I would expect that the ThingHandler does it at some point in time, but calling it from the outside should work as well.

Regarding the actual change of the ThingType, the following line in ThingManager$ThingHandlerCallback.changeThingType(...) changes it in the Thing:

```java
((ThingImpl) thing).setThingTypeUID(thingTypeUID);
```

Please note that the ThingUID is not changed, it stays the same. So if you decide to continue using a 3-segment ThingUID containing the ":generic:" part, it will stay like that. Omitting the middle part would make it less confusing though, but actually is not required.

I added a debug log output at the end of the changeThingType(...) method in order to make this more transparent:

```
15:32:28.550 [main] DEBUG o.e.s.c.thing.internal.ThingManager - Changed ThingType of Thing testBinding::testThing to testBinding:specific. New ThingHandler is org.eclipse.smarthome.core.thing.binding.ChangeThingTypeOSGiTest$SpecificThingHandler@49c2de5c.
```

Also, the PaperUI has been adapted, thanks to @aounhaider1

kaikreuzer commented:
@cdjackson Could you please retest and tell us if it now works for you as expected?

cdjackson commented:
Unfortunately I'm still getting the same exception -:


```
2016-02-02 17:34:08.046 [ERROR] [.c.thing.internal.ThingManager:692  ] - Exception occured while calling thing handler factory 'org.openhab.binding.zwave.internal.ZWaveHandlerFactory@9badfc7': nulljava.util.concurrent.TimeoutException: null
	at java.util.concurrent.FutureTask.get(FutureTask.java:205)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:157)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)
	at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.unregisterHandler(ThingManager.java:684)
	at org.eclipse.smarthome.core.thing.internal.ThingManager.access$14(ThingManager.java:681)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$1.changeThingType(ThingManager.java:238)
	at org.eclipse.smarthome.core.thing.binding.BaseThingHandler.changeThingType(BaseThingHandler.java:540)
	at org.openhab.binding.zwave.handler.ZWaveThingHandler.bridgeHandlerInitialized(ZWaveThingHandler.java:262)
	at org.openhab.binding.zwave.handler.ZWaveThingHandler.initialize(ZWaveThingHandler.java:182)
	at org.eclipse.smarthome.core.thing.internal.ThingManager$7.call(ThingManager.java:626)

```

This is now implemented in the ZWave binding (rather than Bluetooth before) and is called within the initialize method of the thingHandler (although it could also be called at other times). I check to see if the current thingType is the same as it should be based on the zwave database, and change it if it's different (I assume this logic is ok) -:

```
            // Check if we need to change the thing
            if (foundProduct != null && getThing().getThingTypeUID() != foundProduct.getThingTypeUID()) {
                Configuration configuration = editConfiguration();
                changeThingType(foundProduct.getThingTypeUID(), configuration);
            }
```

Again, this doesn't happen every time I run, but it does happen.

I do get the following log entries -:

```
2016-02-02 17:50:54.004 [WARN ] [.c.thing.internal.ThingManager:558  ] - Registering handler for thing 'zwave:device:15292d0d4ca:node2' takes more than 5000ms.
2016-02-02 17:50:54.006 [DEBUG] [.c.thing.internal.ThingManager:257  ] - Changed ThingType of Thing zwave:device:15292d0d4ca:node2 to zwave:fibaro_fgd211_00_000. New ThingHandler is org.openhab.binding.zwave.handler.ZWaveThingHandler@7d5f561c.
2016-02-02 17:50:54.011 [WARN ] [.c.thing.internal.ThingManager:631  ] - Initializing handler for thing 'zwave:device:15292d0d4ca:node2' takes more than 5000ms.
```

So this also indicates some sort of deadlock I guess?

Afterwards, the thing type hasn't changed. If I restart ESH/OH2, then it is definitely still using the original thing type (in this case zwave:device).

cdjackson commented:
Ok - an update. I think it is changing the thing type - it's just not persisted. After the change (which always takes 10 or 15 seconds) it looks like the thing has the new thing type - so this works (although the long delay isn't nice!).  However, if I restart ESH, we are back to the original thing type - I assume this isn't expected?

sjka commented:
@cdjackson It took me a little to dig through all those synchronous call-backs of callbacks of service listeners of callbacks... Yeah, you know it. Anyhow, I discovered a well-hidden "infinite-loop-protection" that actually was a little too generous. Anyhow, I moved the actual thing type migration into another thread, so it gets executed asynchronously in order to break up the endless call chain. I added a test-case for your scenario. My expectation now would be that it runs fast (i.e. takes significantly less than the 5 seconds which the SafeMethodCaller permits.
Also, I added a check that the ManagedThingProvider stores the new ThingTypeUID correctly into the database. So if your HandlerFactory uses the Thing.getThingTypeUID() in order to determine the right handler, my expectation would be that it picks the new Handler type directly after a restart and doesn't go through the old one. 
If still something does not work as expected, please let me know.

cdjackson commented:
Hi Simon,
It seems better - thanks. The thing type is now changed ‘instantly’ :)

The only outstanding issue I think is that the new thing type is not persisted after a restart, so when the binding starts it still has the old (original) thing type and I end up having to perform the discovery again. If I read your email correctly, it should do this right? In some ways this isn’t a major problem, but it takes time, and generates a lot of traffic that slows down initialisation so I think it’s best to persist the new thing type.

FYI… What I am doing is to have a generic thing that all devices when originally discovered into the inbox are set to. Then, in this generic thing handler I perform the ’service’ discovery - i.e. find out what the device really supports and decide the final thing type. I then change it and don’t expect the original generic thing handler to be called again since after a restart the factory will point to the real handler (actually, in zwave I currently only have a single handler for devices and one for the bridge - bluetooth does have multiple handlers though)… I think this is inline with your expectation on use cases, but I thought I’d state it so we’re all on the same page :)

Cheers
Chris

sjka commented:
Just as a quick update: I made some progress today in the sense that I now can reproduce the behavior regarding the restart. Not surprising for you, of course.
I'll now be searching for the root cause of this...

sjka commented:
@cdjackson a couple of headaches later, the database persistence issue should be solved. As well as the initializer-has-been-called-multiple-times-issue, which did not even pop up yet in our discussions...

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java:
```diff
@@ -532,4 +534,12 @@ public void bridgeHandlerDisposed(ThingHandler thingHandler, Bridge bridge) {
+        if (this.callback != null) {
+            this.callback.changeThingType(getThing(), thingTypeUID, configuration);
+        } else {
+            throw new IllegalStateException("Could change thing type because callback is missing");
```
is there a "not" missing?

kaikreuzer commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/ThingHandlerCallback.java:
```diff
@@ -17,15 +19,19 @@
+ *
  * @author Dennis Nobel - Initial contribution
- * @author Stefan Bußweiler - Added new thing status info, added new configuration update info 
+ *         <<<<<<< HEAD
```
this is probably not intended to be in here...

kaikreuzer commented on extensions/ui/org.eclipse.smarthome.ui.paper/web-src/js/controllers.configuration.js:
```diff
@@ -1,623 +1,668 @@
-function getThingTypeUID(thingUID) {
```
It seems you already applied the auto code formatter on the js files - please revert this and save your changes without reformatting, since https://github.com/eclipse/smarthome/pull/1015 is still open.

kaikreuzer commented:
Code lgtm.
@cdjackson Did you already have a chance to see if this finally works for you as expected?

cdjackson commented:
> @cdjackson <https://github.com/cdjackson> Did you already have a chance to see if this finally works for you as expected?
> 
Not yet - I merged the new version earlier tonight so will hopefully have an answer tonight or tomorrow…

cdjackson commented:
Yep - looks good to me...

kaikreuzer commented:
Thanks @cdjackson!
So @sjka: Please address my remaining comments above, squash it and we should be good to merge!

sjka commented:
@kaikreuzer I have incorporated your comments and rebased it. However, the second commit came from @aounhaider1. Do you really want me to squash them?

kaikreuzer commented:
> Do you really want me to squash them?

Yes and simply mention him through an Also-By line.

sjka commented:
okay, done

kaikreuzer commented:
Thanks!
